File: 3670.maximum-product-of-two-integers-with-no-common-bits.java
#
# @lc app=leetcode id=3670 lang=java
#
# [3670] Maximum Product of Two Integers With No Common Bits
#
# @lc code=start
class Solution {
    public long maxProduct(int[] nums) {
        // Map from bitmask to maximum number with that bitmask
        java.util.Map<Integer, Integer> maskToMax = new java.util.HashMap<>();
        for (int num : nums) {
            int mask = num;
            maskToMax.put(mask, Math.max(maskToMax.getOrDefault(mask, 0), num));
        }
        long maxProduct = 0;
        Integer[] masks = maskToMax.keySet().toArray(new Integer[0]);
        int n = masks.length;
        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                if ((masks[i] & masks[j]) == 0) {
                    long prod = (long)maskToMax.get(masks[i]) * maskToMax.get(masks[j]);
                    if (prod > maxProduct) maxProduct = prod;
                }
            }
        }
        return maxProduct;
    }
}
# @lc code=end

File: 3670.maximum-product-of-two-integers-with-no-common-bits.go
#
# @lc app=leetcode id=3670 lang=golang
#
# [3670] Maximum Product of Two Integers With No Common Bits
#

# @lc code=start
func maxProduct(nums []int) int64 {
    maxProduct := int64(0)
    for i := 0; i < len(nums); i++ {
        for j := i + 1; j < len(nums); j++ {
            if nums[i]&nums[j] == 0 { // no common set bits
                product := int64(nums[i]) * int64(nums[j])
                if product > maxProduct {
                    maxProduct = product
                }
            }
        }
    }
    return maxProduct
}
# @lc code=end